src/miniscript/decode.rs

```diff
@@ -34,6 +35,54 @@ fn return_none<T>(_: usize) -> Option<T> {
     None
 }
 
+/// Trait for parsing keys from byte slices
+pub trait ParseableKey: Sized + private::Sealed {
+    /// Parse a key from slice
+    fn from_slice(sl: &[u8]) -> Result<Self, KeyParseError>;
+}
+
+/// Decoding error while parsing keys
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum KeyParseError {
+    /// Bitcoin PublicKey parse error
+    FullKeyParseError(bitcoin::util::key::Error),
+    /// Xonly key parse Error
+    XonlyKeyParseError(bitcoin::secp256k1::Error),
+}
+
+impl ParseableKey for bitcoin::PublicKey {
+    fn from_slice(sl: &[u8]) -> Result<Self, KeyParseError> {
+        bitcoin::PublicKey::from_slice(sl).map_err(KeyParseError::FullKeyParseError)
+    }
+}
+
+impl ParseableKey for bitcoin::schnorr::PublicKey {
+    fn from_slice(sl: &[u8]) -> Result<Self, KeyParseError> {
+        bitcoin::schnorr::PublicKey::from_slice(sl).map_err(KeyParseError::XonlyKeyParseError)
+    }
+}
+
+impl error::Error for KeyParseError {}
+
+impl fmt::Display for KeyParseError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            KeyParseError::FullKeyParseError(e) => write!(f, "FullKey Parse Error {}", e),
+            KeyParseError::XonlyKeyParseError(e) => write!(f, "XonlyKey Parse Error {}", e),
```

Comment:
If you will move the inner error type to `source` (see my comment above) that we should not print inner error here and just use `FullKey parse error` etc